tests/test_app.py

```diff
@@ -1,89 +1,87 @@
+    """
+    with open(file1, "rb") as f1, open(file2, "rb") as f2:
+        while True:
+            b1 = f1.read(chunk_size)
+            b2 = f2.read(chunk_size)
 
-    maskvalues215 = tmpdir / "215/mask_values.txt"
-    maskvalues215expected = expdir / "215/mask_values.txt"
-    assert are_equal(maskvalues215, maskvalues215expected)
+            if b1 != b2:
+                return False
 
-    # Check feature extraction
-    # -----------------------------------------------------------------
-    features214 = tmpdir / "214/2012-08-01_terra_props.csv"
-    features214expected = expdir / "214/2012-08-01_terra_props.csv"
-    assert check_sums(features214, features214expected)
+            if not b1:  # End of file reached
+                return True
 
-    features215 = tmpdir / "215/2012-08-02_terra_props.csv"
-    features215expected = expdir / "215/2012-08-02_terra_props.csv"
-    assert check_sums(features215, features215expected)
 
-    # Check intermediate identification rounds
-    # -----------------------------------------------------------------
-    for doy in ["214", "215"]:
-        id_rounds = sorted(Path(tmpdir / doy).glob("*round*.tif"))
-        expected_rounds = sorted((expdir / doy).glob("*round*.tif"))
+def are_files_identical(file1, file2):
+    return filecmp.cmp(file1, file2, shallow=False)
 
-        for id_round, expected_round in zip(id_rounds, expected_rounds):
-            assert are_equal(id_round, expected_round)
 
+def are_images_identical(image1_path, image2_path):
+    with (
+        rasterio.open(str(image1_path)) as img1,
+        rasterio.open(str(image2_path)) as img2,
+    ):
+        return np.array_equal(img1.read(), img2.read())
 
-def getmaskvalues(path):
-    with open(path, "r") as f:
-        lines = f.readlines()
 
-    mask_values = [float(x) for x in lines[0].split()]
-    return mask_values
+def are_equal(p1, p2):
+    return Path(p1).read_bytes() == Path(p2).read_bytes()
 
 
-def group_files_by_extension(folder_path):
-    grouped_files = defaultdict(list)
-    folder = Path(folder_path)
+def check_sums(p1, p2):
+    s1 = pd.read_csv(p1).to_numpy().sum()
+    s2 = pd.read_csv(p2).to_numpy().sum()
+    return s1 == s2
 
-    for file in folder.iterdir():
-        if file.is_file():
-            extension = file.suffix
-            grouped_files[extension].append(file.name)
 
-    return dict(grouped_files)
+# Check final images
+def _test_output(tmpdir):
+    expdir = Path("tests/expected")
+    # Check all tif images, including final and identification rounds, and csv and txt files
+    # --------------------------------------------------------------------------------------
+    for d in getdirs(tmpdir):
+        for file in d.glob(
+            "*[!.png]"
+        ):  # Exclude png files as they are not created in the original code
+            expected = expdir / file.relative_to(tmpdir)
+            if file.suffix == ".csv":
+                assert check_sums(file, expected)
```

Comment:
Does this check the equality with enough precision? I'm thinking that you might have offsets of +1 and -1 in two of the values, and that would return the same sum, which you might not want to call equality.

Reply:
I think so. These tables are quite big; chances of something like that happening are very small. I tried something like `df1.equal(df2)` but for some reason I am getting differences, perhaps due to threading, of this sort:
```python
for index in differing_indices: print(f"Difference at {index}: df1 = {df1.loc[index]}, df2 = {df2.loc[index]}: {abs(df1.loc[index]-df2
.loc[index])}")
Difference at (21, 'axis_minor_length'): df1 = 18.80043590926547, df2 = 18.800435909265467: 3.552713678800501e-15
Difference at (21, 'perimeter'): df1 = 99.9827560572969, df2 = 99.98275605729692: 2.842170943040401e-14
Difference at (24, 'perimeter'): df1 = 142.50966799187808, df2 = 142.5096679918781: 2.842170943040401e-14
Difference at (28, 'axis_minor_length'): df1 = 20.549389703593025, df2 = 20.54938970359302: 3.552713678800501e-15
Difference at (29, 'orientation'): df1 = 0.6353351033724091, df2 = 0.6353351033724092: 1.1102230246251565e-16
Difference at (31, 'axis_minor_length'): df1 = 31.77893518912647, df2 = 31.778935189126468: 3.552713678800501e-15
Difference at (31, 'perimeter'): df1 = 215.6639969244288, df2 = 215.66399692442883: 2.842170943040401e-14
Difference at (32, 'axis_minor_length'): df1 = 24.933746748747563, df2 = 24.933746748747595: 3.197442310920451e-14
Difference at (34, 'axis_minor_length'): df1 = 28.33918983963977, df2 = 28.33918983963979: 2.1316282072803006e-14

Reply:
It looks like the values are close to within some very tight tolerances, but not identical. You might want to use something like https://pandas.pydata.org/docs/reference/api/pandas.testing.assert_frame_equal.html#pandas.testing.assert_frame_equal instead – 
It defaults to having absolute and relative tolerances for floats which might help. You could always make the precision requirements much more stringent.

Reply:
Do the errors still occur when you're not using the threaded version?

Reply:
Yes, they do. I revised the that call in the test is running serially (setting `max_workers` to `1`) but `df1.equal(df2)` still fails. `pd.testing.assert_frame_equal(df1, df2)` checks out, however. Not sure why this is happening.

Reply:
How close to you expect the values to be? When I look at the numbers in the table above, my first thought is that the calculations in the feature extraction modules can't be that precise. Do they come out exactly the same in each run? I'm not sure how the propagation of machine error into scikit-image or openCV feature calculations works. 

Also, could it be a sorting issue? How unique are the individual rows?

Reply:
> Do they come out exactly the same in each run?

Yes, they come out the same in each run. However, I ran the same job on Oscar and the props table has the same problems. I am thinking it might be due to some difference in environment at a lower level, maybe some linear algebra library or something like that.

I will try running the original code on Oscar and compare the outputs generated in the same computational environment. That way we can make a more apples-to-apples comparison.

Reply:
> When I look at the numbers in the table above, my first thought is that the calculations in the feature extraction modules can't be that precise.

That seems to be the case. See for example one of the props tables in the original repo [here](https://github.com/WilhelmusLab/Segmentation_EB/blob/main/output/2012-08-01_214_terra.csv).

Reply:
> Also, could it be a sorting issue? How unique are the individual rows?

Not sure what you mean by this. Could you elaborate a bit?

Reply:
Sure. Say that two objects have exactly 25 pixels and are close to a square. The perimeter calculation is an approximation. Numpy will spit out numbers with tons of decimal places. But because the perimeter calculation is approximate, there will be uncertainty added at the level of machine error somewhere in the calculation. So you might end up with the two objects having perimeter 20.123456 and 20.123457. (This is also what I was getting at when I said that the calculations can't be that precise -- we don't know the true value to the 15th decimal place.) The last digit might not be stable, so a table sorted by perimeter might look different in one run vs the next.

Of course, if you're also using the centroid in the comparison, they're never going to have centroids that are that close, so the point may be moot.

Reply:
Another thing: is there a rhyme or reason to how the feature extraction tool puts things in order? Are the labels always generated in the same order? That could also lead to differences in the order that floes appear in a table.

Reply:
>Are the labels always generated in the same order?

That seems to be the case. I've run the jobs several times in the same computational environment (Oscar and my local computer) with the same input data and the output is not changing.

Reply:
I see. I am not at all concerned with that level of difference. A tolerance of 1e-5 would be plenty to check that the answers are coming out the same. Assuming these units are in pixels, I don't consider anything past the first decimal to be scientifically meaningful.